File: train_dataset.py
import os
import json
from typing import List, Tuple, Dict, Any, Optional
import numpy as np


class TrainDataset:
    """Utility class to navigate the local training dataset.

    The dataset folder is excluded from version control, so this class
    attempts to hide the exact file layout.  It expects a directory
    structure similar to::

        train/
            s01/
                rep_ann.json
                joints3d_25/
                    squat.json
                    ...
            s02/
                rep_ann.json
                joints3d_25/
                    pushup.json
                    ...

    Each JSON file is expected to contain a key ``"joints3d_25"`` (or a
    similar one) that stores the pose array.  Additional metadata such as
    repetition timings will be returned if present. Repetition timing
    information is primarily loaded from the rep_ann.json file in each
    subject's directory.
    """

    def __init__(self, root: str = "train") -> None:
        self.root = os.fspath(root)

    # ------------------------------------------------------------------
    # listing utilities
    # ------------------------------------------------------------------
    def list_subjects(self) -> List[str]:
        """Return all available subject IDs."""
        if not os.path.isdir(self.root):
            return []
        return sorted(
            d
            for d in os.listdir(self.root)
            if os.path.isdir(os.path.join(self.root, d))
        )

    def list_exercises_for_subject(self, subject: str) -> List[str]:
        """Return exercise names performed by ``subject``."""
        path = os.path.join(self.root, subject, "joints3d_25")
        if not os.path.isdir(path):
            return []
        return sorted(
            os.path.splitext(f)[0]
            for f in os.listdir(path)
            if f.endswith(".json") and os.path.isfile(os.path.join(path, f))
        )

    def list_all_exercises(self) -> List[str]:
        """Return a sorted list of all unique exercises in the dataset."""
        exercises = set()
        for subject in self.list_subjects():
            exercises.update(self.list_exercises_for_subject(subject))
        return sorted(exercises)

    def list_instances(self, exercise: str) -> List[Tuple[str, str]]:
        """Return ``(subject, path)`` tuples for every instance of ``exercise``."""
        instances: List[Tuple[str, str]] = []
        for subject in self.list_subjects():
            path = os.path.join(self.root, subject, "joints3d_25", f"{exercise}.json")
            if os.path.isfile(path):
                instances.append((subject, path))
        return instances

    # ------------------------------------------------------------------
    # loading utilities
    # ------------------------------------------------------------------
    def load_rep_annotations(self, subject: str) -> Optional[Dict[str, List[int]]]:
        """Load the repetition annotations for a given subject.

        Returns a dictionary mapping exercise names to lists of frame numbers
        representing repetition timing annotations, or None if the file doesn't exist.
        """
        rep_ann_path = os.path.join(self.root, subject, "rep_ann.json")
        if not os.path.isfile(rep_ann_path):
            return None

        try:
            with open(rep_ann_path) as fh:
                return json.load(fh)
        except (json.JSONDecodeError, IOError):
            return None

    def load_instance(self, subject: str, exercise: str) -> Dict[str, Any]:
        """Load the full JSON for ``subject`` performing ``exercise``."""
        path = os.path.join(self.root, subject, "joints3d_25", f"{exercise}.json")
        with open(path) as fh:
            data = json.load(fh)

        poses = None
        for key in ("joints3d_25", "joints3d", "poses3d"):
            if key in data:
                poses = np.asarray(data[key])
                break
        if poses is None:
            raise KeyError(f"No pose data found in {path}")

        info = {
            k: v
            for k, v in data.items()
            if k not in {"joints3d_25", "joints3d", "poses3d"}
        }

        # First, try to get timings from rep_ann.json
        timings = None
        rep_annotations = self.load_rep_annotations(subject)
        if rep_annotations and exercise in rep_annotations:
            timings = rep_annotations[exercise]

        # Fall back to timings from the individual exercise file if not found in rep_ann.json
        if timings is None:
            timings = info.get("rep_timings") or info.get("reps") or info.get("timings")

        num_reps = len(timings) if isinstance(timings, list) else info.get("num_reps")

        return {
            "subject": subject,
            "exercise": exercise,
            "poses": poses,
            "timings": timings,
            "num_reps": num_reps,
            "info": info,
        }

    # ------------------------------------------------------------------
    # convenience helpers
    # ------------------------------------------------------------------
    def get_pose_array(self, subject: str, exercise: str) -> np.ndarray:
        """Return only the pose array for ``subject`` and ``exercise``."""
        return self.load_instance(subject, exercise)["poses"]

    def subject_has_exercise(self, subject: str, exercise: str) -> bool:
        """Check whether ``subject`` contains data for ``exercise``."""
        path = os.path.join(self.root, subject, "joints3d_25", f"{exercise}.json")
        return os.path.isfile(path)

    def get_rep_timings(self, subject: str, exercise: str) -> Optional[List[int]]:
        """Get repetition timings for a specific subject and exercise.

        Returns the list of frame numbers representing repetition boundaries,
        or None if no timing information is available.
        """
        instance_data = self.load_instance(subject, exercise)
        return instance_data["timings"]

    def get_rep_segments(
        self, subject: str, exercise: str
    ) -> Optional[List[Tuple[int, int]]]:
        """Get repetition segments as (start, end) frame pairs.

        Returns a list of (start_frame, end_frame) tuples representing individual
        repetitions, or None if no timing information is available.
        """
        timings = self.get_rep_timings(subject, exercise)
        if timings is None or len(timings) < 2:
            return None

        return [(timings[i], timings[i + 1]) for i in range(len(timings) - 1)]

File: src/libs/trainDataset/trainDataset.ts
import fs from "fs";
import path from "path";
import { InstanceData, TrainDataset } from "./trainDatasetTypes";
import { AsyncMethods } from "../utilTypes";

export class TrainDatasetLocal implements TrainDataset {
  root: string;

  constructor(root: string = "train") {
    this.root = path.resolve(root);
  }

  // ------------------------------------------------------------------
  // listing utilities
  // ------------------------------------------------------------------
  listSubjects(): string[] {
    if (!fs.existsSync(this.root)) return [];
    return fs
      .readdirSync(this.root)
      .filter((d) => fs.statSync(path.join(this.root, d)).isDirectory())
      .sort();
  }

  listExercisesForSubject(subject: string): string[] {
    const p = path.join(this.root, subject, "joints3d_25");
    if (!fs.existsSync(p)) return [];
    return fs
      .readdirSync(p)
      .filter(
        (f) => f.endsWith(".json") && fs.statSync(path.join(p, f)).isFile()
      )
      .map((f) => path.parse(f).name)
      .sort();
  }

  listAllExercises(): string[] {
    const ex = new Set<string>();
    for (const s of this.listSubjects()) {
      for (const e of this.listExercisesForSubject(s)) {
        ex.add(e);
      }
    }
    return Array.from(ex).sort();
  }

  listInstances(exercise: string): Array<[string, string]> {
    const instances: Array<[string, string]> = [];
    for (const s of this.listSubjects()) {
      const p = path.join(this.root, s, "joints3d_25", `${exercise}.json`);
      if (fs.existsSync(p)) instances.push([s, p]);
    }
    return instances;
  }

  listCameraIds(subject: string): string[] {
    const p = path.join(this.root, subject, "camera_parameters");
    if (!fs.existsSync(p)) return [];
    return fs
      .readdirSync(p)
      .filter((d) => fs.statSync(path.join(p, d)).isDirectory())
      .sort();
  }

  // ------------------------------------------------------------------
  // loading utilities
  // ------------------------------------------------------------------
  loadRepAnnotations(subject: string): Record<string, number[]> | null {
    const p = path.join(this.root, subject, "rep_ann.json");
    if (!fs.existsSync(p)) return null;
    try {
      const txt = fs.readFileSync(p, "utf8");
      return JSON.parse(txt);
    } catch {
      return null;
    }
  }

  loadInstance(subject: string, exercise: string): InstanceData {
    const p = path.join(this.root, subject, "joints3d_25", `${exercise}.json`);
    const data = JSON.parse(fs.readFileSync(p, "utf8")) as Record<string, any>;

    let poses: [number, number, number][][] | null = null;
    for (const key of ["joints3d_25", "joints3d", "poses3d"]) {
      if (key in data) {
        poses = data[key];
        break;
      }
    }
    if (!poses) throw new Error(`No pose data found in ${p}`);

    const info: Record<string, any> = {};
    for (const [k, v] of Object.entries(data)) {
      if (!["joints3d_25", "joints3d", "poses3d"].includes(k)) {
        info[k] = v;
      }
    }

    let timings: number[] | null = null;
    const rep = this.loadRepAnnotations(subject);
    if (rep && exercise in rep) timings = rep[exercise];
    if (!timings) {
      timings =
        (info["rep_timings"] || info["reps"] || info["timings"]) ?? null;
    }

    const num_reps = Array.isArray(timings) ? timings.length : info["num_reps"];

    return { subject, exercise, poses, timings, num_reps, info };
  }

  // ------------------------------------------------------------------
  // convenience helpers
  // ------------------------------------------------------------------
  getPoseArray(subject: string, exercise: string): number[][][] {
    return this.loadInstance(subject, exercise).poses;
  }

  subjectHasExercise(subject: string, exercise: string): boolean {
    const p = path.join(this.root, subject, "joints3d_25", `${exercise}.json`);
    return fs.existsSync(p);
  }

  getRepTimings(subject: string, exercise: string): number[] | null {
    return this.loadInstance(subject, exercise).timings;
  }

  getRepSegments(
    subject: string,
    exercise: string
  ): Array<[number, number]> | null {
    const t = this.getRepTimings(subject, exercise);
    if (!t || t.length < 2) return null;
    const seg: Array<[number, number]> = [];
    for (let i = 0; i < t.length - 1; i++) {
      seg.push([t[i], t[i + 1]]);
    }
    return seg;
  }

  getVideoBlob(subject: string, exercise: string, cameraId: string): Blob {
    // Try common video file extensions and locations
    const possiblePaths = [
      path.join(this.root, subject, "videos", cameraId, `${exercise}.mp4`),
      path.join(this.root, subject, "videos", cameraId, `${exercise}.webm`),
      path.join(this.root, subject, "videos", cameraId, `${exercise}.mov`),
      path.join(this.root, subject, cameraId, `${exercise}.mp4`),
      path.join(this.root, subject, cameraId, `${exercise}.webm`),
      path.join(this.root, subject, cameraId, `${exercise}.mov`),
    ];

    for (const videoPath of possiblePaths) {
      if (fs.existsSync(videoPath)) {
        const videoBuffer = fs.readFileSync(videoPath);
        const mimeType = this.getMimeTypeFromPath(videoPath);
        return new Blob([videoBuffer], { type: mimeType });
      }
    }

    throw new Error(
      `Video not found for subject: ${subject}, exercise: ${exercise}, camera: ${cameraId}`
    );
  }

  private getMimeTypeFromPath(filePath: string): string {
    const ext = path.extname(filePath).toLowerCase();
    switch (ext) {
      case ".mp4":
        return "video/mp4";
      case ".webm":
        return "video/webm";
      case ".mov":
        return "video/quicktime";
      case ".avi":
        return "video/x-msvideo";
      default:
        return "video/mp4"; // default fallback
    }
  }
}

export class TrainDatasetRemote implements AsyncMethods<TrainDataset> {
  private baseUrl: string;
  private _root: Promise<string>;

  constructor(baseUrl: string = "http://localhost:3001") {
    this.baseUrl = baseUrl;
    this._root = this.fetchRoot();
  }

  private async fetchRoot(): Promise<string> {
    const response = await fetch(`${this.baseUrl}/api/root`);
    if (!response.ok) {
      throw new Error(`Failed to fetch root: ${response.statusText}`);
    }
    const data = await response.json();
    return data.root;
  }

  get root(): Promise<string> {
    return this._root;
  }

  async listSubjects(): Promise<string[]> {
    const response = await fetch(`${this.baseUrl}/api/subjects`);
    if (!response.ok) {
      throw new Error(`Failed to list subjects: ${response.statusText}`);
    }
    return response.json();
  }

  async listExercisesForSubject(subject: string): Promise<string[]> {
    const response = await fetch(
      `${this.baseUrl}/api/exercises/${encodeURIComponent(subject)}`
    );
    if (!response.ok) {
      throw new Error(
        `Failed to list exercises for subject: ${response.statusText}`
      );
    }
    return response.json();
  }

  async listAllExercises(): Promise<string[]> {
    const response = await fetch(`${this.baseUrl}/api/exercises`);
    if (!response.ok) {
      throw new Error(`Failed to list all exercises: ${response.statusText}`);
    }
    return response.json();
  }

  async listInstances(exercise: string): Promise<Array<[string, string]>> {
    const response = await fetch(
      `${this.baseUrl}/api/instances/${encodeURIComponent(exercise)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to list instances: ${response.statusText}`);
    }
    return response.json();
  }

  async listCameraIds(subject: string): Promise<string[]> {
    const response = await fetch(
      `${this.baseUrl}/api/camera-ids/${encodeURIComponent(subject)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to list camera IDs: ${response.statusText}`);
    }
    return response.json();
  }

  async loadRepAnnotations(
    subject: string
  ): Promise<Record<string, number[]> | null> {
    const response = await fetch(
      `${this.baseUrl}/api/rep-annotations/${encodeURIComponent(subject)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to load rep annotations: ${response.statusText}`);
    }
    return response.json();
  }

  async loadInstance(subject: string, exercise: string): Promise<InstanceData> {
    const response = await fetch(
      `${this.baseUrl}/api/instance/${encodeURIComponent(
        subject
      )}/${encodeURIComponent(exercise)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to load instance: ${response.statusText}`);
    }
    return response.json();
  }

  async getPoseArray(subject: string, exercise: string): Promise<number[][][]> {
    const response = await fetch(
      `${this.baseUrl}/api/poses/${encodeURIComponent(
        subject
      )}/${encodeURIComponent(exercise)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to get pose array: ${response.statusText}`);
    }
    return response.json();
  }

  async subjectHasExercise(
    subject: string,
    exercise: string
  ): Promise<boolean> {
    const response = await fetch(
      `${this.baseUrl}/api/has-exercise/${encodeURIComponent(
        subject
      )}/${encodeURIComponent(exercise)}`
    );
    if (!response.ok) {
      throw new Error(
        `Failed to check if subject has exercise: ${response.statusText}`
      );
    }
    return response.json();
  }

  async getRepTimings(
    subject: string,
    exercise: string
  ): Promise<number[] | null> {
    const response = await fetch(
      `${this.baseUrl}/api/rep-timings/${encodeURIComponent(
        subject
      )}/${encodeURIComponent(exercise)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to get rep timings: ${response.statusText}`);
    }
    return response.json();
  }

  async getRepSegments(
    subject: string,
    exercise: string
  ): Promise<Array<[number, number]> | null> {
    const response = await fetch(
      `${this.baseUrl}/api/rep-segments/${encodeURIComponent(
        subject
      )}/${encodeURIComponent(exercise)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to get rep segments: ${response.statusText}`);
    }
    return response.json();
  }

  async getVideoBlob(
    subject: string,
    exercise: string,
    cameraId: string
  ): Promise<Blob> {
    const response = await fetch(
      `${this.baseUrl}/api/video-blob/${encodeURIComponent(
        subject
      )}/${encodeURIComponent(exercise)}/${encodeURIComponent(cameraId)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to get video blob: ${response.statusText}`);
    }
    return response.blob();
  }
}
